### PR TITLE
ref zf2#7498 timeout exceptions for curl timeout errors

### PR DESCRIPTION
**Pull request by makv**
related to https://github.com/zendframework/zf2/issues/7498

### CODE REVIEW

File: src/Client/Adapter/Curl.php [hunk 1]
```diff
@@ -425,7 +425,18 @@ public function write($method, $uri, $httpVersion = 1.1, $headers = [], $body =
         $request .= $body;
 
         if (empty($this->response)) {
-            throw new AdapterException\RuntimeException("Error in cURL request: " . curl_error($this->curl));
+            $curlErrorNumber = curl_errno($this->curl);
+            $errorMessage = "Error in cURL request: " . curl_error($this->curl);
+            if ($curlErrorNumber == CURLE_OPERATION_TIMEOUTED) {
+                preg_match('/Connection timed out/', $errorMessage, $matches);
+                if (isset($matches[0])) {
```
Comment (Maks3w) [inline src/Client/Adapter/Curl.php]: use the value returned from preg_match instead

**Comment by Maks3w** [comment]
Need tests for success scenarios with the same settings

File: src/Client/Adapter/Exception/ConnectTimeoutException.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace Zend\Http\Client\Adapter\Exception;
+
+/**
+ */
```
Comment (Maks3w) [inline src/Client/Adapter/Exception/ConnectTimeoutException.php]: Please add a description detailing why this is different from other timeout exceptions

File: test/Client/CurlTest.php [hunk 3]
```diff
@@ -397,4 +397,33 @@ public function testSetCurlOptPostFields()
...
+        $timeout = 1;
+        $timeoutInMilliseconds = $timeout * 1000 + 1;
+        $this->client->setUri($this->baseuri . "testTimeout.php?timeout=$timeout");
+        $adapter = new Adapter\Curl();
+        $adapter->setOptions([
+            'curloptions' => [CURLOPT_TIMEOUT => $timeout]
+        ]);
+        $this->client->setAdapter($adapter);
+        $this->client->setMethod('GET');
+        $this->setExpectedException('Zend\Http\Client\Adapter\Exception\TimeoutException', "Error in cURL request: Operation timed out after $timeoutInMilliseconds milliseconds with 0 bytes received");
```
Comment (Maks3w) [inline test/Client/CurlTest.php]: Use PHP 5.5 ::class and chop down the params following PSR-2 CS guidelines

File: test/Client/CurlTest.php [hunk 4]
```diff
@@ -397,4 +397,33 @@ public function testSetCurlOptPostFields()
...
+    {
+        $timeoutInMilliseconds = 1001;
+        $this->client->setUri('http://10.0.0.0');
+        $adapter = new Adapter\Curl();
+        $adapter->setOptions([
+            'curloptions' => [CURLOPT_CONNECTTIMEOUT_MS => $timeoutInMilliseconds - 1]
+        ]);
+        $this->client->setAdapter($adapter);
+        $this->client->setMethod('GET');
+        $this->setExpectedException('Zend\Http\Client\Adapter\Exception\ConnectTimeoutException', "Error in cURL request: Connection timed out after $timeoutInMilliseconds milliseconds");
```
Comment (Maks3w) [inline test/Client/CurlTest.php]: Use PHP 5.5 ::class and chop down the params following PSR-2 CS guidelines

**Comment by ezimuel** [comment]
I close this PR due to inactivity.

**Pull request closed without merging**